pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -128,23 +132,33 @@ func (s *ServiceEntryStore) workloadEntryHandler(old, curr model.Config, event m
 	}
 	log.Debugf("Handle event %s for workload entry %s in namespace %s", event, curr.Name, curr.Namespace)
 	instances := []*model.ServiceInstance{}
+	instancesDelete := []*model.ServiceInstance{}
+
 	for _, se := range entries {
 		workloadLabels := labels.Collection{wle.Labels}
 		if !workloadLabels.IsSupersetOf(se.entry.WorkloadSelector.Labels) {
-			// Not a match, skip this one
-			continue
+			if oldWle != nil {
+				oldWorkloadLabels := labels.Collection{oldWle.Labels}
+				if oldWorkloadLabels.IsSupersetOf(se.entry.WorkloadSelector.Labels) {
+					instance := convertWorkloadEntryToServiceInstances(oldWle, se.services, se.entry)
+					instancesDelete = append(instancesDelete, instance...)
+				}
+			}
+		} else {
+			instance := convertWorkloadEntryToServiceInstances(wle, se.services, se.entry)
+			instances = append(instances, instance...)
 		}
-		instance := convertWorkloadEntryToServiceInstances(wle, se.services, se.entry)
-		instances = append(instances, instance...)
 	}
 
+	s.deleteExistingInstances(key, instancesDelete)
+
 	if event != model.EventDelete {
 		s.updateExistingInstances(key, instances)
```

Comment:
My understanding is `update` will set the current instances to `instances`, which will not have the mismatched labels because we `continue` on L135. So why is the special delete needed?

Reply:
The `instances` only include the service entries that can select the workload entry by labels.

If you change the workload entry labels and make it not match from match, the `instances` will be empty.

So I created a new `instance` list that can select old workload entry and can not select the new one. And we need remove the workload entry.

You can reproduce this issue based on this issue #25678, it's very easy to reproduce it.